#167123181 User Login On Frontend

BrianSerem opened the pull request:
## Description ##
- Enables users to login to LandVille. Heroku pipeline [here](https://landville-frontend-pr-11.herokuapp.com/login):

## Type of change ##
Please select the relevant option
- [ ] Bug fix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes
- [ ] End to end test
- [ ] Integration test
- [x] unit test

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

## Related Pivotal Tracker stories ##
[#167123181](https://www.pivotaltracker.com/story/show/167123181)

BrianSerem committed:
feature(login) : users login to frontend

- Create login form, slider, header and social login.
- Create http service that hits backend
- Write tests

[Starts #167246164]

BrianSerem committed:
feat(login) : modify tests that were failing

-create http service
-create config file
-rewrite tests

[#167123181]

BrianSerem committed:
feature(login) : user login

-rewrite tests to pass

[#167123181]

BrianSerem committed:
feature(login) : create custom errors

-create new error handlers.
-verify if tests running.

[#167123181]

CryceTruly commented on src/app/pages/login/login-form/login-form.component.spec.ts:
```diff
@@ -0,0 +1,132 @@
+        done();
+    }, 3000);
+});
+  it('an invalid form passed', () => {
```
@BrianSerem, thanks for the good work, However it's customary that 'it' is followed by 'should' in your test descriptions, How you used it on line 99 can be a great good example.

BrianSerem replied:
Yes sure, I have renamed the test description to contain the word 'should'

CryceTruly replied:
Good job

CryceTruly commented on src/app/pages/login/login-form/login-form.component.ts:
```diff
@@ -0,0 +1,69 @@
+      this.notification = 'Login was succesful';
+      localStorage.setItem('token', response.data.user.token)
+
+      this.router.navigate([''])
```
@BrianSerem, I believe you might need to take advantage of router guards to prevent authenticated users from accessing the login component when they are logged in. As is now, after a successful login a backward navigation will still show users the login component.

BrianSerem replied:
I believe we earlier spoke about this. We need a new story to handle guarding the routes we currently have. Will gladly take it up after I am done with this feature.

CryceTruly replied:
great @BrianSerem

CryceTruly commented on src/app/pages/login/login.component.scss:
```diff
@@ -1,9 +1,41 @@
 
-    }
-  }
+@media screen and (max-width: 1024px){
```
Good work Brian on implementing navigation for small screens, However, I realize on small screens and in device landscape mode, the margin on the left is bigger, while there is no margin on the right. You might need to look into it

BrianSerem replied:
I have worked on this, I remove the background image on the login page for smaller devices, I additionally ensured that the login form components fits symmetrically on the screen.

CryceTruly commented on src/app/pages/login/login-form/login-form.component.ts:
```diff
@@ -0,0 +1,69 @@
+  templateUrl: './login-form.component.html',
+  styleUrls: ['./login-form.component.scss']
+})
+export class LoginFormComponent implements OnInit {
```
@BrianSerem, I realize the login page has the generic LandVille title, I think we should customize the metadata of the page especially the title, to something like `LandVille Login' or something more descriptive. We could achieve this by passing this metadata in the Login Route configurations in our Routing Module.

CryceTruly commented on src/app/pages/login/login-form/login-form.component.ts:
```diff
@@ -0,0 +1,69 @@
+    })
+    
+  }
+  setErrorTimeout(){
```
@BrianSerem, I realize that when a user gets an error or when their credentials are invalid,all input fields are cleared, I think it is good we keep their previous data in the fields such that they can edit rather than retyping afresh.

BrianSerem replied:
I have removed the loginForm.reset() method to ensure that form data persist after an error.

CryceTruly reviewed (changes requested):
Great work Brian, please check through the reviews I left down below.

CryceTruly commented on src/app/helpers/spies.ts:
```diff
@@ -0,0 +1,21 @@
+export const  toastServiceSpy = createSpyObj('ToastrService', [
+    'success',
+    'error'
+]);
```
@BrianSerem, We eliminated the blank line here, but we should keep it

BrianSerem replied:
I have added an extra line here.

BrianSerem committed:
feature(login) : work on feedback

-make login page more responsive
-remove form.reset method after error

BrianSerem committed:
feat(login):work on feedback

-install and activate ngx-spinner
-work on mobile responsiveness
-create custom titles for components

[#167123181]

CryceTruly reviewed (approved):
LGTM
